runtime/eventring_consumer.c

```diff
@@ -0,0 +1,693 @@
+  if (Is_some(tmp_callback)) {
+    params[0] = Val_long(domain_id);
+    params[1] = caml_copy_int64(timestamp);
+    params[2] = Val_long(counter);
+    params[3] = Val_long(val);
+
+    caml_callbackN(Some_val(tmp_callback), 4, params);
+  }
+
+  CAMLdrop;
+  return 1;
+}
+
+static int ml_alloc(int domain_id, void *callback_data, uint64_t timestamp,
+                     uint64_t *sz) {
+  CAMLparam0();
+  CAMLlocal4(tmp_callback, ts_val, misc_val, callbacks_root);
+
+  callbacks_root = *((value *)callback_data);
+
+  tmp_callback = Field(callbacks_root, 3); /* ev_alloc */
+  if (Is_some(tmp_callback)) {
+    int i;
+
+    ts_val = caml_copy_int64(timestamp);
+    misc_val = caml_alloc(EVENTRING_NUM_ALLOC_BUCKETS, 0);
+
+    for (i = 0; i < EVENTRING_NUM_ALLOC_BUCKETS; i++) {
+      Store_field(misc_val, i, Val_long(sz[i]));
+    }
+
+    caml_callback3(Some_val(tmp_callback), Val_long(domain_id), ts_val,
+                   misc_val);
+  }
+
+  CAMLdrop;
+  return 1;
+}
+
+static int ml_lifecycle(int domain_id, void *callback_data, int64_t timestamp,
+                         ev_lifecycle lifecycle, int64_t data) {
+  CAMLparam0();
+  CAMLlocal2(tmp_callback, callbacks_root);
+  CAMLlocalN(params, 4);
+
+  callbacks_root = *((value *)callback_data);
+
+  tmp_callback = Field(callbacks_root, 4); /* ev_lifecycle */
+  if (Is_some(tmp_callback)) {
+    params[0] = Val_long(domain_id);
+    params[1] = caml_copy_int64(timestamp);
+    params[2] = Val_long(lifecycle);
+    if (data != 0) {
+      params[3] = caml_alloc(1, 0);
+      Store_field(params[3], 0, Val_long(data));
+    } else {
+      params[3] = Val_none;
+    }
+
+    caml_callbackN(Some_val(tmp_callback), 4, params);
+  }
+
+  CAMLdrop;
+  return 1;
+}
+
+static int ml_lost_events(int domain_id, void *callback_data, int lost_words) {
+  CAMLparam0();
+  CAMLlocal2(tmp_callback, callbacks_root);
+
+  callbacks_root = *((value *)callback_data);
+
+  tmp_callback = Field(callbacks_root, 5); /* lost_events */
+
+  if (Is_some(tmp_callback)) {
+    caml_callback2(Some_val(tmp_callback), Val_long(domain_id),
+                   Val_long(lost_words));
+  }
+
+  CAMLdrop;
+  return 1;
+}
+
+static struct custom_operations cursor_operations = {
+    "eventring.cursor",         finalise_cursor,
+    custom_compare_default,     custom_hash_default,
+    custom_serialize_default,   custom_deserialize_default,
+    custom_compare_ext_default, custom_fixed_length_default};
+
+CAMLprim value caml_ml_eventring_create_cursor(value path_pid_option) {
+  CAMLparam1(path_pid_option);
+  CAMLlocal1(wrapper);
+  struct caml_eventring_cursor *cursor;
+  int pid;
+  const char_os* path;
+  eventring_error res;
+
+  wrapper = caml_alloc_custom(&cursor_operations,
+                              sizeof(struct caml_eventring_cursor *), 0, 1);
+
+  Cursor_val(wrapper) = NULL;
+
+  if (Is_some(path_pid_option)) {
+    const char* path_u8 = String_val(Field(path_pid_option, 0));
```

Comment:
I think this needs the extra `Some_val`to access the allocated option.

```suggestion
    const char* path_u8 = String_val(Field(Some_val(path_pid_option), 0));
```